egs/wsj/s5/steps/libs/nnet3/xconfig/self_layer.py

```diff
@@ -0,0 +1,218 @@
+        # the SelfAttentionComponent for the computation of linear combination of frames.
+        # The tricky part here is that the softmax computation are done in the SelfAttentionComponent
+        # as the weights of each head are arranged in columns of the CuMatrixBase
+        # We need to transpose the weights to row vector, then call ApplySoftMaxPerRow(), 
+        # then transpose the weights back to column vector.
+
+        da_dim = self._affine_dim
+        self_repair_scale = 1e-05
+        target_rms = 1.0
+        affine_options = 'max-change=0.75'
+
+        configs = []
+        # The first affine node.
+        configs.append(
+            'component name={0}.first_affine type=NaturalGradientAffineComponent'
+            ' input-dim={1} output-dim={2} {3}'
+            ''.format(self.name, input_dim, da_dim, affine_options))
+
+        configs.append(
+            'component-node name={0}.first_affine component={0}.first_affine input={1}'
+            ''.format(self.name, input_desc))
+
+        # The ReLU node.
+        configs.append(
+             'component name={0}.relu type=RectifiedLinearComponent dim={1}'
+             ' self-repair-scale={2}'
+             ''.format(self.name, da_dim, self_repair_scale))
+
+        configs.append(
+             'component-node name={0}.relu'
+             ' component={0}.relu input={0}.first_affine'
+             ''.format(self.name))
+
+        # The batchnorm node.
+        configs.append(
+             'component name={0}.batchnorm type=BatchNormComponent dim={1}'
+             ' target-rms={2}'
+             ''.format(self.name, da_dim, target_rms))
+
+        configs.append(
+             'component-node name={0}.batchnorm'
+             ' component={0}.batchnorm input={0}.relu'
+             ''.format(self.name))
+
+        # The second affine node.
+        affine_options = 'param-stddev=0.04472135955 bias-stddev=1.0 bias-mean=0.0 max-change=0.75 l2-regularize=0.0'
+        configs.append(
+            'component name={0}.second_affine type=NaturalGradientAffineComponent'
+            ' input-dim={1} output-dim={2} {3}'
+            ''.format(self.name, da_dim, self._num_heads, affine_options))
+
+        configs.append(
+            'component-node name={0}.second_affine component={0}.second_affine input={0}.batchnorm'
+            ''.format(self.name))
+
+
+        # The statistic extraction node.
+        configs.append(
+            'component name={name}-extraction-{lc}-{rc} '
+            'type=StatisticsExtractionComponent input-dim={dim} '
+            'input-period={input_period} output-period={output_period} '
+            'include-variance={var} '.format(
+                name=self.name, lc=self._left_context, rc=self._right_context,
+                dim=input_dim, input_period=self._input_period,
+                output_period=self._stats_period,
+                var='true' if self._output_stddev else 'false'))
```

Comment:
@david-ryan-snyder, you should probably review this too.
This uses StatisticsExtractionComponent but without the StaticsPoolingComponent, which is odd;
it's, in effect, pooling short groups of frames before the self-attention.  
Also, this layer seems to be a bit against the spirit of a layer; it has a couple of Affine+Relu+Batchnorm before the poling and attention, so should probably be split into two layers.
@vimalmanohar, if you have specific ideas to improve this, please comment too.

Reply:
@tomkocse can correct me if I'm wrong, but I don't think this PR is meant to be merged in it's current state. It's related to some experiments we're doing. 

I think there's more work to be done on this, both to establish it's value as a layer and to improve the code (again, @tomkocse, correct me if I'm wrong).

Reply:
@david-ryan-snyder Yes, I agree what you said, let's wait for more results from further experiments.